### PR TITLE
refactor: build row groups from record batch stream

### PR DESCRIPTION
**Pull request by ShiKaiWi**
# Which issue does this PR close?

Refactor for #370

# Rationale for this change
 Previous implementation for building row groups from record batch is a little bit complex, and this pr will do some refactoring work for that.

# What changes are included in this PR?
- Refactor the `partition_record_batch`;

# Are there any user-facing changes?
None.

# How does this change test
Existing ut.

### CODE REVIEW

File: analytic_engine/src/sst/parquet/builder.rs [hunk 1]
```diff
@@ -67,80 +64,72 @@ impl RecordBytesReader {
...
-        let mut remaining = Vec::with_capacity(pending_record_batch.len());
-        while let Some(batch) = pending_record_batch.pop_front() {
-            remaining.push(batch);
-        }
-        if !remaining.is_empty() {
-            self.partitioned_record_batch.push(remaining);
+        Ok(())
+    }
+
+    /// Fetch an integrate row group from the `self.record_stream`.
```
Comment (jiacai2050) [inline analytic_engine/src/sst/parquet/builder.rs:80]: What does `integrate row group` means?

File: analytic_engine/src/sst/parquet/builder.rs [hunk 2]
```diff
@@ -67,80 +64,72 @@ impl RecordBytesReader {
...
+        let mut curr_row_group = vec![];
+        // Used to record the number of remaining rows to fill `curr_row_group`.
+        let mut remaining = self.num_rows_per_row_group;
+
+        // Keep filling `curr_row_group` until `remaining` is zero.
+        while remaining > 0 {
+            // Use the `prev_record_batch` to fill `curr_row_group` if possible.
+            if let Some(v) = prev_record_batch {
+                let total_rows = v.num_rows();
+                if total_rows <= remaining {
```
Comment (jiacai2050) [inline analytic_engine/src/sst/parquet/builder.rs:98]: Should be `total_rows < remaining`?

File: analytic_engine/src/sst/parquet/builder.rs [hunk 3]
```diff
@@ -67,80 +64,72 @@ impl RecordBytesReader {
...
+                    // The whole record batch is part of the `curr_row_group`, and let's feed it
+                    // into `curr_row_group`.
+                    curr_row_group.push(prev_record_batch.take().unwrap());
+                    remaining -= total_rows;
+                } else {
+                    // Only part of the record batch fills `curr_row_group`, and let's fill the
```
Comment (jiacai2050) [inline analytic_engine/src/sst/parquet/builder.rs:104]: How about
> Only first `remaining` record batch belongs to `curr_row_group`, the rest should be put to `prev_record_batch` for next row_group

**Comment by jiacai2050** [comment]
The general implementation is more clearer than before, good job!

File: analytic_engine/src/sst/parquet/builder.rs (hunk 1 above)
Comment (ShiKaiWi) [reply analytic_engine/src/sst/parquet/builder.rs:80]: The word `integrate` may be a wrong usage here. Maybe `integral` or `complete` is better.

Actually, it means the number of the row group is `num_rows_per_group`.

File: analytic_engine/src/sst/parquet/builder.rs (hunk 2 above)
Comment (ShiKaiWi) [reply analytic_engine/src/sst/parquet/builder.rs:98]: I guess it should be `<=` here.

File: analytic_engine/src/sst/parquet/builder.rs (hunk 3 above)
Comment (ShiKaiWi) [reply analytic_engine/src/sst/parquet/builder.rs:104]: Ok. It's nicer, I will put it like:
> Only first `remaining` rows of the record batch belongs to `curr_row_group`, the rest should be put to `prev_record_batch` for next row group.

**Review by jiacai2050 — APPROVED** [review]
LGTM

**Pull request merged**